proxy/agent.go

```diff
@@ -214,18 +214,26 @@ func (a *agent) Run(ctx context.Context) {
 			// cleanup for the epoch
 			a.proxy.Cleanup(status.epoch)
 
-			// schedule a retry for a transient error and skip aborts
-			if status.err != nil && status.err != errAbort && !reflect.DeepEqual(a.desiredConfig, a.currentConfig) {
-				if a.retry.budget > 0 {
-					delayDuration := a.retry.InitialInterval * (1 << uint(a.retry.MaxRetries-a.retry.budget))
-					restart := time.Now().Add(delayDuration)
-					a.retry.restart = &restart
-					a.retry.budget = a.retry.budget - 1
-					glog.V(2).Infof("Updated retry delay to %v, budget to %d", delayDuration, a.retry.budget)
+			// schedule a retry for an error.
+			// the current config might be out of date from here since its proxy might have been aborted.
+			// the current config will change on abort, hence retrying prior to abort will not progress.
+			// that means that aborted envoy might need to re-schedule a retry if it was not already scheduled.
+			if status.err != nil {
+				// skip retrying twice by checking retry restart delay
+				if a.retry.restart == nil {
+					if a.retry.budget > 0 {
+						delayDuration := a.retry.InitialInterval * (1 << uint(a.retry.MaxRetries-a.retry.budget))
+						restart := time.Now().Add(delayDuration)
+						a.retry.restart = &restart
+						a.retry.budget = a.retry.budget - 1
```

Comment:
is there no -- operator in go?
a.retry.budget--

Reply:
Just a suggestion how about the following less nested code?

	if status.err == nil {
		continue
	}
	// skip retrying twice by checking retry restart delay
	if a.retry.restart == nil {
		glog.V(2).Infof("Epoch %d: restart already scheduled", status.epoch)
        	continue
	}

	if a.retry.budget > 0 {
		delayDuration := a.retry.InitialInterval * (1 << uint(a.retry.MaxRetries-a.retry.budget))
		restart := time.Now().Add(delayDuration)
		a.retry.restart = &restart
		a.retry.budget--
		glog.V(2).Infof("Epoch %d: set retry delay to %v, budget to %d", status.epoch, delayDuration, a.retry.budget)
	} else {
		glog.Error("Permanent error: budget exhausted trying to fulfill the desired configuration")
		a.proxy.Panic(a.desiredConfig)
		return
	}

Reply:
`--` is fine, just doesn't fit with the rest of the code style. i tried not to reorganize code too much, but your suggestion looks better i think.